recipes_source/mobile_perf.rst

```diff
@@ -0,0 +1,223 @@
+of other hardware backends such as GPU, DSP, and NPU.
+
+
+Model preparation
+-----------------
+
+Next recipes you can take (offline) while preparing the model
+to have an optimized model that will probably have shorter execution time
+(higher performance, lower latency) on the mobile device.
+
+
+Setup
+_____
+First we need to installed pytorch using conda or pip with version at least 1.5.0.
+
+::
+
+   conda install pytorch torchvision -c pytorch
+
+or
+
+::
+
+   pip install torch torchvision
+
+Code your model:
+
+::
+
+  import torch
+  from torch.utils.mobile_optimizer import optimize_for_mobile
+
+  class AnnotatedConvBnReLUModel(torch.nn.Module):
+      def __init__(self):
+          super(AnnotatedConvBnReLUModel, self).__init__()
+          self.conv = torch.nn.Conv2d(3, 5, 3, bias=False).to(dtype=torch.float)
+          self.bn = torch.nn.BatchNorm2d(5).to(dtype=torch.float)
+          self.relu = torch.nn.ReLU(inplace=True)
+          self.quant = torch.quantization.QuantStub()
+          self.dequant = torch.quantization.DeQuantStub()
+
+      def forward(self, x):
+          x = self.quant(x)
+          x = self.conv(x)
+          x = self.bn(x)
+          x = self.relu(x)
+          x = self.dequant(x)
+          return x
+
+  model = AnnotatedConvBnReLUModel()
+
+
+``torch.quantization.QuantStub`` and ``torch.quantization.DeQuantStub()`` are no-op stubs, which will be used for quantization step.
+
+
+1. Fuse operators using ``torch.quantization.fuse_modules``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Do not be confused that fuse_modules is in the quantization package.
+It works for all ``torcn.nn.Module``.
+
+``torch.quantization.fuse_modules`` fuses a list of modules into a single module.
+It fuses only the following sequence of modules:
+
+- Convolution, Batch normalization
+- Convolution, Batch normalization, Relu
+- Convolution, Relu
+- Linear, Relu
+
+This script will fuse Convolution, Batch Normalization and Relu in previously declared model.
+
+::
+
+  torch.quantization.fuse_modules(model, [['conv', 'bn', 'relu']], inplace=True)
+
+
+2. Quantize your model
+~~~~~~~~~~~~~~~~~~~~~~
+
+You can find more about PyTorch quantization in
+`the dedicated tutorial <https://pytorch.org/blog/introduction-to-quantization-on-pytorch/>`_.
+
+Quantization of the model not only moves computation to int8,
+but also reduces the size of your model on a disk.
+That size reduction helps to reduce disk read operations during the first load of the model and decreases the amount of RAM.
+Both of those resources can be crucial for the performance of mobile applications.
+This code does quantization, using stub for model calibration function, you can find more about it `here <https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html#post-training-static-quantization>`__.
+
+::
+
+  model.qconfig = torch.quantization.get_default_qconfig('qnnpack')
+  torch.quantization.prepare(model)
```

Comment:
`prepare()` has `inplace=False` by default, and the same goes for `convert()`, so this whole method is a no-op except for setting `model.qconfig`.

We either need to do:
```
model = torch.quantization.prepare(model)
# calibration
return torch.quantization.convert(model)
```

or:
```
torch.quantization.prepare(model, inplace=True)
# calibration
torch.quantization.convert(model, inplace=True)
```